eppoclient/evalbandits.go

```diff
@@ -0,0 +1,234 @@
+package eppoclient
+
+import (
+	"math"
+	"sort"
+	"strconv"
+)
+
+type ContextAttributes struct {
+	Numeric     map[string]float64
+	Categorical map[string]string
+}
+
+// Tries to map generic attributes to ContextAttributes depending on attribute types.
+// - Integer and float types are mapped to numeric attributes.
+// - Strings and bools are mapped to categorical attributes.
+// - Rest of types are silently dropped.
+func InferContextAttributes(attrs map[string]interface{}) ContextAttributes {
+	result := ContextAttributes{
+		Numeric:     map[string]float64{},
+		Categorical: map[string]string{},
+	}
+	for key, value := range attrs {
+		switch value := value.(type) {
+		case int:
+			result.Numeric[key] = float64(value)
+		case int8:
+			result.Numeric[key] = float64(value)
+		case int16:
+			result.Numeric[key] = float64(value)
+		case int32:
+			result.Numeric[key] = float64(value)
+		case int64:
+			result.Numeric[key] = float64(value)
+		case uint:
+			result.Numeric[key] = float64(value)
+		case uint8:
+			result.Numeric[key] = float64(value)
+		case uint16:
+			result.Numeric[key] = float64(value)
+		case uint32:
+			result.Numeric[key] = float64(value)
+		case uint64:
+			result.Numeric[key] = float64(value)
+		case float32:
+			result.Numeric[key] = float64(value)
+		case float64:
+			result.Numeric[key] = value
+		case string:
+			result.Categorical[key] = value
+		case bool:
+			result.Categorical[key] = strconv.FormatBool(value)
+		}
+	}
+	return result
+}
+
+func (self ContextAttributes) toGenericAttributes() Attributes {
+	result := make(Attributes)
+	for key, value := range self.Numeric {
+		result[key] = value
+	}
+	for key, value := range self.Categorical {
+		result[key] = value
+	}
+	return result
+}
+
+type banditEvaluationContext struct {
+	flagKey           string
+	subjectKey        string
+	subjectAttributes ContextAttributes
+	actions           map[string]ContextAttributes
+}
+
+type banditEvaluationDetails struct {
+	flagKey           string
+	subjectKey        string
+	subjectAttributes ContextAttributes
+	result            *banditEvaluationResult
+}
+type banditEvaluationResult struct {
+	actionKey        string
+	actionAttributes ContextAttributes
+	actionScore      float64
+	actionWeight     float64
+	gamma            float64
+	optimalityGap    float64
+}
+
+type action struct {
+	key        string
+	attributes ContextAttributes
+}
+
+func (model *banditModelData) evaluate(ctx banditEvaluationContext) banditEvaluationDetails {
+	// THere's currently no way to change totalShards in bandit evaluation.
+	var totalShards int64 = 10_000
+
+	nActions := len(ctx.actions)
+	if nActions == 0 {
+		return banditEvaluationDetails{
+			flagKey:           ctx.flagKey,
+			subjectKey:        ctx.subjectKey,
+			subjectAttributes: ctx.subjectAttributes,
+			result:            nil,
+		}
+	}
```

Comment:
per my other comment, I wonder if we should prevent getting here upstream and just return default value. That way the top-level function always returns something usable.

Reply:
I have updated the upstream function. Now `banditEvaluationDetails.result` can never be `nil`, so I inlined it.